src/main/java/io/jenkins/plugins/ml/IPythonKernelInterpreter.java

```diff
@@ -0,0 +1,129 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.jenkins.plugins.ml;
+
+import org.apache.zeppelin.interpreter.*;
+import org.apache.zeppelin.python.IPythonInterpreter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Properties;
+
+public class IPythonKernelInterpreter implements KernelInterpreter  {
+
+    /**
+     * Our logger
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(IPythonKernelInterpreter.class);
+    private final String serverGatewayAddress;
+    private final long iPythonLaunchTimeout;
+    private final long maxResult;
+
+    private LazyOpenInterpreter interpreter;
+    private InterpreterResultMessage interpreterResultMessage ;
+
+    /**
+     * @param userConfig - user's configuration including server address etc.
+     */
+    public IPythonKernelInterpreter(IPythonUserConfig userConfig) {
+        this.serverGatewayAddress = userConfig.getServerGatewayAddress();
+        this.iPythonLaunchTimeout = userConfig.getIPythonLaunchTimeout();
+        this.maxResult = userConfig.getMaxResult();
+
+        // properties for the interpreter
+        Properties properties = new Properties();
+        properties.setProperty("zeppelin.python.maxResult", String.valueOf(maxResult));
+        properties.setProperty("zeppelin.python.gatewayserver_address", serverGatewayAddress);
+        properties.setProperty("zeppelin.ipython.launch.timeout", String.valueOf(iPythonLaunchTimeout));
+
+        // Initiate a Lazy interpreter
+        interpreter = new LazyOpenInterpreter(new IPythonInterpreter(properties));
+
+    }
+
+    /**
+     * IPython will be connected and interpreted through the object of this class.
+     * @param code - python code to be executed
+     * @return the result of the interpreted code
+     */
+    @Override
+    public InterpreterResultMessage interpretCode(String code) throws IOException {
+        InterpreterResult result;
+        InterpreterContext context = getInterpreterContext();
+        try {
+            result = interpreter.interpret(code, context);
+            LOGGER.info(result.code().toString());
+            int sizeOfResult = context.out.toInterpreterResultMessage().size();
+
+            // Context output will depend on the code to be interpreted
+            if( sizeOfResult != 0){
+                interpreterResultMessage = context.out.toInterpreterResultMessage().get(0);
+            }else{
+                // Contains non-output lines
+                interpreterResultMessage = new InterpreterResultMessage(InterpreterResult.Type.TEXT,"");
+            }
+
+        } catch (InterpreterException e) {
+            LOGGER.error("Interpretation failed ", e);
+        }
+        return interpreterResultMessage;
```

Comment:
I think it would be better to throw the `InterpreterException` in the `IPythonKernelInterpreter`, instead of handling errors in the methods, and returning `null` or void in the methods of this class.

Otherwise, in this case you will need to check the return type of this method in your `Builder` or some other extension point, to tell whether it was executed correctly or not.

Reply:
agreed +1